go/vt/vtadmin/cache/cache.go

```diff
@@ -0,0 +1,255 @@
+		cfg.BackfillEnqueueWaitTime = DefaultBackfillEnqueueWaitTime
+	}
+
+	if cfg.BackfillRequestTTL <= 0 {
+		log.Warningf("BackfillRequestTTL (%v) must be positive, defaulting to %v", cfg.BackfillRequestTTL, DefaultBackfillRequestTTL)
+		cfg.BackfillRequestTTL = DefaultBackfillRequestTTL
+	}
+
+	c := &Cache[Key, Value]{
+		cache:     cache.New(cfg.DefaultExpiration, cfg.CleanupInterval),
+		lastFill:  map[string]time.Time{},
+		backfills: make(chan *backfillRequest[Key], cfg.BackfillQueueSize),
+		fillFunc:  fillFunc,
+		cfg:       cfg,
+	}
+
+	c.ctx, c.cancel = context.WithCancel(context.Background())
+
+	c.wg.Add(1)
+	go c.backfill() // TODO: consider allowing N backfill threads to run, configurable
+
+	return c
+}
+
+// Add adds a (key, value) to the cache directly, following the semantics of
+// (github.com/patrickmn/go-cache).Cache.Add.
+func (c *Cache[Key, Value]) Add(key Key, val Value, d time.Duration) error {
+	return c.add(key.Key(), val, d)
+}
+
+func (c *Cache[Key, Value]) add(key string, val Value, d time.Duration) error {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	// Record the time we last cached this key, to check against
+	c.lastFill[key] = time.Now().UTC()
+	// Then cache the actual value.
+	return c.cache.Add(key, val, d)
+}
+
+// Get returns the Value stored for the key, if present in the cache. If the key
+// is not cached, the zero value for the given type is returned, along with a
+// boolean to indicated presence/absence.
+func (c *Cache[Key, Value]) Get(key Key) (Value, bool) {
+	v, exp, ok := c.cache.GetWithExpiration(key.Key())
+	if !ok || (!exp.IsZero() && exp.Before(time.Now())) {
+		var zero Value
+		return zero, false
+	}
+
+	return v.(Value), ok
+}
+
+// EnqueueBackfill submits a request to the backfill queue.
+func (c *Cache[Key, Value]) EnqueueBackfill(k Key) bool {
+	req := &backfillRequest[Key]{
+		k:           k,
+		requestedAt: time.Now().UTC(),
+	}
+
+	select {
+	case c.backfills <- req:
+		return true
+	case <-time.After(c.cfg.BackfillEnqueueWaitTime):
+		return false
+	}
+}
+
+// Close closes the backfill goroutine, effectively rendering this cache
+// unusable for further background fills.
+func (c *Cache[Key, Value]) Close() error {
+	c.cancel()
+	c.wg.Wait()
+	return nil
+}
+
+type backfillRequest[Key Keyer] struct {
+	k           Key
+	requestedAt time.Time
+}
+
+func (c *Cache[Key, Value]) backfill() {
+	defer c.wg.Done()
+
+	for {
+		var req *backfillRequest[Key]
+		select {
+		case <-c.ctx.Done():
+			return
+		case req = <-c.backfills:
+		}
+
+		if req.requestedAt.Add(c.cfg.BackfillRequestTTL).Before(time.Now()) {
+			// We took too long to get to this request, per config options.
+			log.Warningf("backfill for %s requested at %s; discarding due to exceeding TTL (%s)", req.k.Key(), req.requestedAt, c.cfg.BackfillRequestTTL)
+			continue
+		}
+
+		key := req.k.Key()
+
+		c.m.Lock()
+		if t, ok := c.lastFill[key]; ok {
```

Comment:
I see you've decided not to clean-up `lastFill`, which probably makes sense for most growth patterns. I would comment this behavior explicitly.